Django 6 support

In NixOS we run all dependency tests to make sure the software we package works even if we loosen some version constraints.

For Django 6 some tests broke. The majority I already tried to fix in https://github.com/django-extensions/django-extensions/pull/1977 but there is another one broken which I think is because of the example application but I am not sure:

```
pytest flags: -m pytest --ignore-glob=tests/test_dumpscript.py
============================= test session starts ==============================
platform linux -- Python 3.13.12, pytest-9.0.2, pluggy-1.6.0
django: version: 6.0.3, settings: tests.testapp.settings (from ini)
rootdir: /build/source
configfile: pyproject.toml
plugins: Faker-40.1.2, cov-nixpkgs-stub-1.0.0, django-4.11.1
collected 523 items / 1 skipped                                                

tests/auth/test_mixins.py ..                                             [  0%]
tests/db/fields/test_uniq_field_mixin.py .......                         [  1%]
tests/db/fields/test_uniq_field_mixin_compat.py .......                  [  3%]
tests/jobs/daily/test_cache_cleanup.py ..                                [  3%]
tests/management/commands/shell_plus_tests/test_collision_resolver.py .. [  3%]
........................                                                 [  8%]
tests/management/commands/shell_plus_tests/test_import_subclasses.py ... [  8%]
........                                                                 [ 10%]
tests/management/commands/shell_plus_tests/test_shell_plus.py ..         [ 10%]
tests/management/commands/test_admin_generator.py ....                   [ 11%]
tests/management/commands/test_clear_cache.py ......                     [ 12%]
tests/management/commands/test_create_command.py ....                    [ 13%]
tests/management/commands/test_create_jobs.py ....                       [ 14%]
tests/management/commands/test_create_template_tags.py ...               [ 14%]
tests/management/commands/test_delete_squashed_migrations.py XXXXXxxXxx  [ 16%]
tests/management/commands/test_describe_form.py ....                     [ 17%]
tests/management/commands/test_drop_test_database.py .............       [ 20%]
tests/management/commands/test_export_emails.py .........                [ 21%]
tests/management/commands/test_generate_secret_key.py .                  [ 21%]
tests/management/commands/test_graph_models.py ..........                [ 23%]
tests/management/commands/test_list_signals.py F                         [ 24%]
tests/management/commands/test_managestate.py ...............            [ 26%]
tests/management/commands/test_print_user_for_session.py ......          [ 28%]
tests/management/commands/test_reset_db.py ............                  [ 30%]
tests/management/commands/test_reset_schema.py ....                      [ 31%]
tests/management/commands/test_runjob.py .........                       [ 32%]
tests/management/commands/test_runserver_plus.py .                       [ 33%]
tests/management/commands/test_set_default_site.py ...........           [ 35%]
tests/management/commands/test_set_fake_emails.py .........              [ 36%]
tests/management/commands/test_set_fake_passwords.py ....                [ 37%]
tests/management/commands/test_show_permissions.py .......               [ 39%]
tests/management/commands/test_show_urls.py ..........                   [ 40%]
tests/management/commands/test_sqlcreate.py .......                      [ 42%]
tests/management/commands/test_sqldsn.py .............                   [ 44%]
tests/management/commands/test_sync_s3.py ...........                    [ 46%]
tests/management/commands/test_syncdata.py ........                      [ 48%]
tests/management/commands/test_unreferenced_files.py ....                [ 49%]
tests/management/commands/test_update_permissions.py ....                [ 49%]
tests/management/commands/test_validate_templates.py ....                [ 50%]
tests/management/test_email_notifications.py ....                        [ 51%]
tests/templatetags/test_highlighting.py ....                             [ 52%]
tests/templatetags/test_indent_text.py ....                              [ 52%]
tests/templatetags/test_syntax_color.py ........                         [ 54%]
tests/test_admin_filter.py .....                                         [ 55%]
tests/test_admin_widgets.py .                                            [ 55%]
tests/test_autoslug_fields.py ..........................                 [ 60%]
tests/test_clean_pyc.py ...                                              [ 61%]
tests/test_compat.py ...                                                 [ 61%]
tests/test_compile_pyc.py ..                                             [ 62%]
tests/test_find_template.py ..                                           [ 62%]
tests/test_json_field.py ...........                                     [ 64%]
tests/test_logging_filters.py ...                                        [ 65%]
tests/test_management_command.py .......................                 [ 69%]
tests/test_models.py ......                                              [ 70%]
tests/test_modificationdatetime_fields.py ...                            [ 71%]
tests/test_parse_mysql_cnf.py ...                                        [ 71%]
tests/test_randomchar_field.py ........F...                              [ 74%]
tests/test_runscript.py ....................................             [ 81%]
tests/test_shortuuid_field.py ....                                       [ 81%]
tests/test_sqldiff.py ...sss...                                          [ 83%]
tests/test_template_rendering.py ..                                      [ 83%]
tests/test_templatetags.py ...                                           [ 84%]
tests/test_timestamped_model.py ..                                       [ 84%]
tests/test_validators.py ......                                          [ 86%]
django_extensions/management/shells.py .                                 [ 86%]
django_extensions/templatetags/widont.py ..                              [ 86%]
tests/management/commands/shell_plus_tests/test_shell_plus.py ........   [ 88%]
tests/management/commands/test_generate_password.py ..                   [ 88%]
tests/management/commands/test_graph_models.py ....                      [ 89%]
tests/management/commands/test_mail_debug.py .                           [ 89%]
tests/management/commands/test_notes.py ....                             [ 90%]
tests/management/commands/test_print_settings.py .......                 [ 91%]
tests/management/commands/test_raise_test_exception.py .                 [ 91%]
tests/management/commands/test_set_fake_emails.py .                      [ 91%]
tests/management/commands/test_set_fake_passwords.py .                   [ 92%]
tests/management/commands/test_show_template_tags.py ...                 [ 92%]
tests/management/commands/test_validate_templates.py .                   [ 92%]
tests/management/test_modelviz.py ....                                   [ 93%]
tests/test_color.py ..                                                   [ 94%]
tests/test_runserver.py ....................                             [ 97%]
tests/test_validators.py ...........                                     [100%]

=================================== FAILURES ===================================
________________ ListSignalsTests.test_should_print_all_signals ________________

self = <tests.management.commands.test_list_signals.ListSignalsTests testMethod=test_should_print_all_signals>

        def test_should_print_all_signals(self):
            expected_result = """django.contrib.sites.models.Site (site)
        pre_delete
            django.contrib.sites.models.clear_site_cache #
        pre_save
            django.contrib.sites.models.clear_site_cache #
    tests.testapp.models.HasOwnerModel (has owner model)
        pre_save
            tests.testapp.models.dummy_handler #
    """
    
            call_command("list_signals", stdout=self.out)
    
            # Strip line numbers to make the test less brittle
            out = re.sub(r"(?<=#)\d+", "", self.out.getvalue(), re.M)
>           self.assertIn(expected_result, out)
E           AssertionError: 'django.contrib.sites.models.Site (site)\n    pre_delete\n        django.contrib.sites.models.clear_site_cache #\n    pre_save\n        django.contrib.sites.models.clear_site_cache #\ntests.testapp.models.HasOwnerModel (has owner model)\n    pre_save\n        tests.testapp.models.dummy_handler #\n' not found in 'django.contrib.sites.models.Site (site)\n    pre_delete\n        django.contrib.sites.models.clear_site_cache # (async)\n    pre_save\n        django.contrib.sites.models.clear_site_cache # (async)\ntests.testapp.models.HasOwnerModel (has owner model)\n    pre_save\n        tests.testapp.models.dummy_handler # (async)\n'

tests/management/commands/test_list_signals.py:30: AssertionError
_____________ RandomCharFieldTest.testRandomCharTestModelDuplicate _____________

self = <tests.test_randomchar_field.RandomCharFieldTest testMethod=testRandomCharTestModelDuplicate>

    def testRandomCharTestModelDuplicate(self):
        m = RandomCharTestModelUnique()
        m.save()
        with mock.patch(
            "django_extensions.db.fields.RandomCharField.random_char_generator"
        ) as func:
            func.return_value = iter([m.random_char_field, "aaa"])
            m = RandomCharTestModelUnique()
>           m.save()

tests/test_randomchar_field.py:88: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:874: in save
    self.save_base(
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:966: in save_base
    updated = self._save_table(
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:1169: in _save_table
    results = self._do_insert(
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:1220: in _do_insert
    return manager._insert(
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/manager.py:87: in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/query.py:1918: in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/sql/compiler.py:1924: in execute_sql
    for sql, params in self.as_sql():
                       ^^^^^^^^^^^^^
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/sql/compiler.py:1813: in as_sql
    field_prepare(field_pre_save(obj)) for obj in self.query.objs
                  ^^^^^^^^^^^^^^^^^^^
/nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/sql/compiler.py:1751: in pre_save_val
    return field.pre_save(obj, add=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
django_extensions/db/fields/__init__.py:418: in pre_save
    return self.find_unique(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <django_extensions.db.fields.RandomCharField: random_char_field>
model_instance = <RandomCharTestModelUnique: RandomCharTestModelUnique object (None)>
field = <django_extensions.db.fields.RandomCharField: random_char_field>
iterator = <list_iterator object at 0x7ffff2124b80>, args = ()
queryset = <QuerySet []>, kwargs = {}

    def find_unique(self, model_instance, field, iterator, *args):
        # exclude the current model instance from the queryset used in finding
        # next valid hash
        queryset = self.get_queryset(model_instance.__class__, field)
        if model_instance.pk:
            queryset = queryset.exclude(pk=model_instance.pk)
    
        # form a kwarg dict used to implement any unique_together constraints
        kwargs = {}
        for params in model_instance._meta.unique_together:
            if self.attname in params:
                for param in params:
                    kwargs[param] = getattr(model_instance, param, None)
    
        # for support django 2.2+
        query = Q()
        constraints = getattr(model_instance._meta, "constraints", None)
        if constraints:
            unique_constraints = filter(
                lambda c: isinstance(c, UniqueConstraint), constraints
            )
            for unique_constraint in unique_constraints:
                if self.attname in unique_constraint.fields:
                    condition = {
                        field: getattr(model_instance, field, None)
                        for field in unique_constraint.fields
                        if field != self.attname
                    }
                    query &= Q(**condition)
    
>       new = next(iterator)
              ^^^^^^^^^^^^^^
E       StopIteration

django_extensions/db/fields/__init__.py:88: StopIteration
=============================== warnings summary ===============================
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandEror_if_more_than_one_migration_matches_to_given_arg
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandEror_if_squashed_migration_not_found
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandEror_if_squashed_migration_not_foundee
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandError_if_app_does_not_have_migrations
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandError_when_database_does_not_exist
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_delete_squashed_migrations
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_delete_squashed_migrations_if_interactive_mode_is_set_to_False
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_not_delete_anything
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_not_delete_files_for_given_squashed_migration
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'testapp_with_appconfig.titlemodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsExceptionsTests::test_should_raise_CommandEror_if_squashed_migration_not_found
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_delete_squashed_migrations_if_interactive_mode_is_set_to_False
tests/management/commands/test_delete_squashed_migrations.py::DeleteSquashedMigrationsTests::test_should_not_delete_files_for_given_squashed_migration
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'testapp_with_appconfig.namemodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/management/commands/test_describe_form.py::DescribeFormTests::test_should_print_form_definition_for_TestModel_with_non_editable_field
tests/management/commands/test_describe_form.py::DescribeFormTests::test_should_print_form_with_fields_for_TestModel
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'testapp.basemodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/management/commands/test_describe_form.py::DescribeFormTests::test_should_print_form_definition_for_TestModel_with_non_editable_field
tests/management/commands/test_describe_form.py::DescribeFormTests::test_should_print_form_with_fields_for_TestModel
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'testapp.noneditablemodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/management/commands/test_drop_test_database.py: 11 warnings
tests/management/commands/test_reset_db.py: 3 warnings
tests/management/commands/test_reset_schema.py: 1 warning
tests/management/commands/test_sqlcreate.py: 6 warnings
tests/management/commands/test_sqldsn.py: 13 warnings
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/test/utils.py:455: UserWarning: Overriding setting DATABASES can lead to unexpected behavior.
    with self as context:

tests/management/commands/test_list_signals.py::ListSignalsTests::test_should_print_all_signals
  /build/source/tests/management/commands/test_list_signals.py:29: DeprecationWarning: 'count' is passed as positional argument
    out = re.sub(r"(?<=#)\d+", "", self.out.getvalue(), re.M)

tests/management/commands/test_reset_db.py::ResetDbSqlite3Tests::test_should_cancel_reset_db_if_input_is_different_than_yes
tests/management/commands/test_reset_db.py::ResetDbMysqlTests::test_should_cancel_reset_db_if_input_is_different_than_yes
tests/management/commands/test_reset_db.py::ResetDbPostgresqlTests::test_should_cancel_reset_db_if_input_is_different_than_yes
tests/management/commands/test_reset_schema.py::ResetSchemaTests::test_should_cancel_reset_schema_and_print_info_if_input_is_different_than_yes
  /nix/store/m1fw8l8y9ycxh5dzispbb7cwl6rra14l-python3-3.13.12/lib/python3.13/unittest/case.py:116: UserWarning: Overriding setting DATABASES can lead to unexpected behavior.
    result = enter(cm)

tests/management/commands/test_syncdata.py::SyncDataTests::test_should_delete_old_objects_and_load_data_from_json_fixture
tests/management/commands/test_syncdata.py::SyncDataTests::test_should_keep_old_objects_and_load_data_from_json_fixture
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/fields/__init__.py:1670: RuntimeWarning: DateTimeField User.date_joined received a naive datetime (2019-01-19 13:37:26) while time zone support is active.
    warnings.warn(

tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_should_reload_permission_only_for_specified_apps
tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_should_update_permission_if_name_changed
tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_works
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'django_extensions.permmodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_should_reload_permission_only_for_specified_apps
tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_should_update_permission_if_name_changed
tests/management/commands/test_update_permissions.py::UpdatePermissionsTests::test_works
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/base.py:388: RuntimeWarning: Model 'testapp.testmodel' was already registered. Reloading models is not advised as it can lead to inconsistencies, most notably with related models.
    new_class._meta.apps.register_model(new_class._meta.app_label, new_class)

tests/test_modificationdatetime_fields.py::ModificationDatetimeFieldTest::test_disabled_update_modification_field
  /nix/store/3yckabw2isx8gl0ci9950935kchb3cg4-python3.13-django-6.0.3/lib/python3.13/site-packages/django/db/models/fields/__init__.py:1670: RuntimeWarning: DateTimeField DisabledUpdateModelModificationDateTimeField.modified received a naive datetime (2026-03-08 12:59:10.713652) while time zone support is active.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED tests/management/commands/test_list_signals.py::ListSignalsTests::test_should_print_all_signals - AssertionError: 'django.contrib.sites.models.Site (site)\n    pre_delete\n ...
FAILED tests/test_randomchar_field.py::RandomCharFieldTest::testRandomCharTestModelDuplicate - StopIteration
= 2 failed, 508 passed, 4 skipped, 4 xfailed, 6 xpassed, 64 warnings in 22.52s =

```

--- a/django_extensions/db/fields/__init__.py
+++ b/django_extensions/db/fields/__init__.py
@@ -86,7 +86,11 @@ class UniqueFieldMixin:
 
         new = next(iterator)
         kwargs[self.attname] = new
-        while not new or queryset.filter(query, **kwargs):
+        while True:
+            matching = queryset.filter(query, **kwargs)
+            has_match = matching.exists() if hasattr(matching, "exists") else bool(matching)
+            if new and not has_match:
+                break
             new = next(iterator)
             kwargs[self.attname] = new
         setattr(model_instance, self.attname, new)
@@ -388,10 +392,14 @@ class RandomCharField(UniqueFieldMixin, CharField):
         return False
 
     def pre_save(self, model_instance, add):
-        if (not add or self.keep_default) and getattr(
-            model_instance, self.attname
-        ) != "":
-            return getattr(model_instance, self.attname)
+        current_value = getattr(model_instance, self.attname)
+        # Django 6 may call pre_save multiple times for inserts; if we've already
+        # populated the field value, reuse it instead of regenerating.
+        if current_value not in ("", None):
+            return current_value
+
+        if (not add or self.keep_default) and current_value != "":
+            return current_value
 
         population = ""
         if self.include_alpha:

--- a/django_extensions/management/commands/list_signals.py
+++ b/django_extensions/management/commands/list_signals.py
@@ -52,7 +52,9 @@ class Command(BaseCommand):
         for signal in signals:
             signal_name = SIGNAL_NAMES.get(signal, "unknown")
             for receiver in signal.receivers:
-                if django.VERSION >= (5, 0):
+                if django.VERSION >= (6, 0):
+                    lookup, receiver, _sender, is_async = receiver
+                elif django.VERSION >= (5, 0):
                     lookup, receiver, is_async = receiver
                 else:
                     lookup, receiver = receiver

--- a/tests/test_management_command.py
+++ b/tests/test_management_command.py
@@ -3,6 +3,7 @@ from unittest import mock
 import logging
 import importlib
 
+import django
 from django.core.management import (
     call_command,
     find_commands,
@@ -421,7 +422,10 @@ class ListModelInfoTests(TestCase):
             stdout=out,
         )
         self.output = out.getvalue()
-        self.assertIn("id - AutoField", self.output)
+        if django.VERSION >= (6, 0):
+            self.assertIn("id - BigAutoField", self.output)
+        else:
+            self.assertIn("id - AutoField", self.output)
         self.assertIn("char_field - CharField", self.output)
         self.assertIn("integer_field - IntegerField", self.output)
         self.assertIn("foreign_key_field - ForeignKey", self.output)

--- a/tests/testapp/models.py
+++ b/tests/testapp/models.py
@@ -1,3 +1,4 @@
+import django
 from django.db import models
 from django.contrib.auth import get_user_model
 from django.db.models import UniqueConstraint
@@ -178,7 +179,11 @@ class PostWithUniqField(models.Model):
                 fields=("common_field", "uniq_field"), name="unique_common_uniq_pair"
             ),
             models.CheckConstraint(
-                check=~models.Q(common_field=models.F("another_common_field")),
+                **(
+                    {"condition": ~models.Q(common_field=models.F("another_common_field"))}
+                    if django.VERSION >= (5, 2)
+                    else {"check": ~models.Q(common_field=models.F("another_common_field"))}
+                ),
                 name="common_and_another_common_differ",
             ),
         ]

